### PR TITLE
Fix State Pruning

### PR DESCRIPTION
**Pull request by nisdas**
This fixes state pruning so that only states with slots below the last finalized slot are removed from the db.

### CODE REVIEW

File: beacon-chain/db/state.go [hunk 1]
```diff
@@ -156,8 +156,9 @@ func (db *BeaconDB) SaveJustifiedState(beaconState *pb.BeaconState) error {
 
 // SaveFinalizedState saves the last finalized state in the db.
 func (db *BeaconDB) SaveFinalizedState(beaconState *pb.BeaconState) error {
+	finalizedSlot := beaconState.FinalizedEpoch * params.BeaconConfig().SlotsPerEpoch
```
Comment (terenc3t) [inline beacon-chain/db/state.go]: we can use `helper.SlotToEpoch(beaconState.FinalizedEpoch)` here

File: beacon-chain/db/state.go [hunk 2]
```diff
@@ -213,9 +214,9 @@ func (db *BeaconDB) SaveCurrentAndFinalizedState(beaconState *pb.BeaconState) er
 	}
 
 	db.currentState = currentState
-
+	finalizedSlot := beaconState.FinalizedEpoch * params.BeaconConfig().SlotsPerEpoch
```
Comment (terenc3t) [inline beacon-chain/db/state.go]: `helper.slotToEpoch(beaconState.FinalizedEpoch)`

File: beacon-chain/db/state.go [hunk 3]
```diff
@@ -339,18 +340,22 @@ func (db *BeaconDB) GenesisTime(ctx context.Context) (time.Time, error) {
 	return genesisTime, nil
 }
 
-func (db *BeaconDB) deleteHistoricalStates() error {
+func (db *BeaconDB) deleteHistoricalStates(slot uint64) error {
 	return db.update(func(tx *bolt.Tx) error {
 		histState := tx.Bucket(histStateBucket)
 		chainInfo := tx.Bucket(chainInfoBucket)
 		hsCursor := histState.Cursor()
+		slotDiff := slot - params.BeaconConfig().GenesisSlot
```
Comment (terenc3t) [inline beacon-chain/db/state.go]:
```suggestion
		slotsSinceGenesis := slot - params.BeaconConfig().GenesisSlot
```

File: beacon-chain/db/state.go (hunk 1 above)
Comment (nisdas) [reply beacon-chain/db/state.go]: I didn't wan't to bring the core package into here, imo the db package should be as self-contained as possible

Comment (rauljordan) [reply beacon-chain/db/state.go]: I think we should have those helpers in the shared package, we also need them in the validator client btw

Comment (rauljordan) [reply beacon-chain/db/state.go]: It can be a separate PR

File: beacon-chain/db/state.go (hunk 2 above)
Comment (prestonvanloon) [reply beacon-chain/db/state.go]: +1

File: beacon-chain/db/state.go (hunk 1 above)
Comment (prestonvanloon) [reply beacon-chain/db/state.go]: ag

**Pull request merged**